---
title: "Transcriptional Dynamics of Milk Produiction in Bovine Mammary Glands"
author: "Benji Lamp"
date: "4/24/24"
output: 
  html_document: 
    toc: true
    toc_float: true
    theme: united
    highlights: tango
runtime: shiny
---

# Introduction

It is understood that the mammary gland in cattle undergoes extensive morphological and physiological transformations to facilitate the production of milk. These changes are regulated at the genetic and epigenetic level, with specific patterns of expression driving maturation of the gland. This study aims to uncover the temporal expression patterns of mRNA in bovine mammary glands by performing timewise differential expression of sequencing data, then clustering based on temporal expression pattern.

# Setup {.tabset}

### Libraries

```{r, message=FALSE, warning=FALSE}
library(DESeq2)
library(dplyr)
library(AnnotationDbi)
library(org.Bt.eg.db)
library(sva)
library(readxl)
library(tidyr)
library(knitr)
library(kableExtra)
library(ggplot2)
library(plotly)
library(ggrepel)

```

### Data

The raw count files can be downloaded directly from Github. The data was loaded using the read.delim function from the base package, setting the Ensembl ids as rownames. The first column, which was redundant (gene-ids), was removed.

```{r}
file_path <- 'https://raw.githubusercontent.com/blamp23/Cattle-lac-proj/main/Raw_data/Raw_Cattle_Counts.tabular'
df <- read.delim(file_path, header = TRUE, row.names = 1)
df <- df[-1]
```

# Filtering/loading

## Metadata

Sample metadata was compiled to reflect the stages of pregnancy as well as the batches associated with strandedness. The metadata table includes animal identifier, stage of pregnancy, and batch for each sample.

```{r, echo=FALSE}
samples <- data.frame(
  row.names = c("v_468", "v_502", "v_504", "v_507", "v_509", "v_598", "v_610", 
                "mp_468", "mp_502", "mp_504", "mp_507", "mp_509", "mp_598", "mp_610",
                "lp_468", "lp_502", "lp_504", "lp_509", "lp_507", "lp_598", "lp_610",
                "el_468", "el_502", "el_504", "el_507", "el_509", "el_598", "el_610",
                "l_468", "l_502", "l_504", "l_507", "l_509", "l_598", "l_610"),
  condition = c(rep("v", 7), rep("mp", 7), rep("lp", 7), rep("el", 7), rep("l", 7)),
  animal = c(468, 502, 504, 507, 509, 598, 610,
             468, 502, 504, 507, 509, 598, 610,
             468, 502, 504, 509, 507, 598, 610,
             468, 502, 504, 507, 509, 598, 610,
             468, 502, 504, 507, 509, 598, 610),
  batch = c(1, 1, 1, 1, 1, 1, 1, 
            0, 1, 0, 1, 0, 0, 0,
            0, 1, 0, 0, 1, 0, 0,
            0, 1, 0, 1, 0, 0, 0, 
            1, 1, 1, 1, 1, 1, 1 )
)
```

Before finalizing the raw counts, it's necessary to adjust the column names of the dataframe to simplify data manipulation. The format is the condition followed by the sample identifier (e.g., XX_123). The format mimics the row names of the metadata dataframe. Additionally, it is important to note that the samples from Early Lactation 610 and Late Pregnancy 610 were switched prior to sequencing and must be corrected before proceeding with the analysis.

```{r, echo=F, results='asis', message=FALSE}
new_colnames <- c(row.names(samples))
colnames(df) <- new_colnames

colnames(df)[28] <- "lp_610"  # lp_610 and el_610 are swapped samples
colnames(df)[21] <- "el_610"

el_610 <- which(names(df) == "el_610")
lp_610 <- which(names(df) == "lp_610")

# Create a new vector of column names with the positions of these two columns swapped
new_order <- names(df)
new_order[c(el_610, lp_610)] <- new_order[c(lp_610, el_610)]

# Reorder the columns in the DataFrame
df <- df[, new_order]

samples <- samples[colnames(df), ]



# Custom CSS to enlarge table captions
cat("
<style>
table caption {
  font-size: 16pt;
  font-weight: bold;
}
</style>
")

# Assuming 'samples' is another data frame you want to display
kable(samples, caption = "Sample Data") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = FALSE)

```

## Filtering

Each gene is evaluated within each lactation stage based on its count in the 7 samples. A gene needs to have a count above the threshold of 3 in at least 4 of the 7 samples per time-point to be flagged for inclusion. If a gene has high counts in at least 1/5 of the timepoints, it will be included in the study.

```{r}
# Filter Method A ##############################################################
# originaly 27607 genes
# filtered to 17344 genes
# Step 1: Splitting Data by Condition
conditions <- unique(samples$condition)
split_data <- lapply(conditions, function(cond) {
  df[, samples$condition == cond]
})

# Step 2: Define Filtering Function
count_high_expressions <- function(data) {
  apply(data, 1, function(gene) {
    sum(gene >= 3) >= 4  # Returns TRUE if at least 4 out of 7 samples have counts â‰¥ 3
  })
}
high_exp_counts <- lapply(split_data, count_high_expressions)

# Step 3: Aggregate Results for inclusion
gene_inclusion_filter <- rowSums(do.call(cbind, high_exp_counts)) >= 1 # If 1/5 timepoints meet crit, gene is included

# Step 4: Filter Genes
df <- df[gene_inclusion_filter, ]
```

# Norm

## ComBatSeq

Batch effects that resulted from strandedness which were seen to skew the analysis results in previous principal component analysis, were corrected using the ComBat_seq function from the sva package. This function adjusts for known batch variables, specified by the batch parameter in the metadata, and prevents over fitting by taking into account the experimental design (time course variables).

```{r}
count_matrix <- as.matrix(df)

cdf <- ComBat_seq(count_matrix,
                  batch = samples$batch,
                  group = samples$condition)
```

## DESeq2

```{r, message=FALSE, warning=FALSE}
dds <- DESeqDataSetFromMatrix(
  countData = cdf,
  colData = samples,
  design = ~ condition
)



```

### PCA

```{r, message=FALSE, warning=FALSE, echo=F}
vsd <- vst(dds)
# Get the PCA data
pcaData <- plotPCA(vsd, intgroup=c("condition"), returnData=TRUE)

# Extract the percent variance captured by each principal component
percentVar <- round(100 * attr(pcaData, "percentVar"))

# Plot with ggplot2
ggplot(pcaData, aes(x=PC1, y=PC2, color=condition)) +
  geom_point(size=3) +
  geom_text_repel(aes(label=row.names(pcaData))) +  # using ggrepel to prevent overlap
  stat_ellipse(aes(group=condition), level=0.85) +  # adding ellipses around each condition group
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  theme_bw() +
  ggtitle("PCA Plot")

```

# Timecourse

## Averages DF

```{r, echo=FALSE, message=FALSE}
dds <- estimateSizeFactors(dds)
normalized_counts_df <- as.data.frame(counts(dds, normalized=TRUE))
col_names <- rep(c("v", "mp", "lp", "el", "l"), each = 7)
names(normalized_counts_df) <- col_names

# Prepare an empty data frame for storing averages, with correct dimensions and column names
averages_df <- data.frame(matrix(ncol = length(unique(col_names)), nrow = nrow(normalized_counts_df)))
names(averages_df) <- unique(col_names)
rownames(averages_df) <- rownames(normalized_counts_df)

# Loop through each time-point to calculate and store row means
for (timepoint in unique(col_names)) {
  # Identify columns belonging to the current time-point
  cols <- which(col_names == timepoint)
  # Calculate and store
  averages_df[[timepoint]] <- rowMeans(normalized_counts_df[, cols], na.rm = TRUE)
}

```

```{r, echo=FALSE, message=FALSE, results='asis'}
hadf <- head(averages_df)
# Custom CSS to enlarge table captions
cat("
<style>
table caption {
  font-size: 16pt;
  font-weight: bold;
}
</style>
")

# Assuming 'samples' is another data frame you want to display
kable(hadf, caption = "Averages") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = FALSE)
```

## Pairwise Comparison

```{r}
contrasts <- list(
  c("condition", "l", "v"),
  #
  c("condition", "el", "v"),
  c("condition", "l", "mp"),
  #
  c("condition", "lp", "v"),
  c("condition", "el", "mp"),
  c("condition", "l", "lp"),
  #
  c("condition", "mp", "v"),
  c("condition", "lp", "mp"),
  c("condition", "el", "lp"),
  c("condition", "l", "el")
)

contrast_results <- list()

# DE for pairs in contrasts
for (i in seq_along(contrasts)) {
  contrast = contrasts[[i]]
  res <- results(dds, contrast=contrast)
  
  # Determine if the mean expression increases or decreases
  condition1_mean <- averages_df[[contrast[3]]]  # Mean for the first condition
  condition2_mean <- averages_df[[contrast[2]]]  # Mean for the second condition
  expressionDirection <- ifelse(condition1_mean < condition2_mean, "I", "D")
  
  # Hypothesis test, FTR null assign 'S', Rej null assign I/D
  res$expressionChange <- ifelse(res$padj > 0.05, "S", expressionDirection)
  
  contrast_results[[paste(contrast[3], "to", contrast[2])]] <- res
}
```

## Motif Structuring

```{r}
combined_results <- data.frame(gene=rownames(contrast_results[[1]]))
# Loop through each contrast result to combine them
for (i in seq_along(contrast_results)) {
  contrast_name <- names(contrast_results)[i]  # Get the name of the current contrast
  # Create a temporary dataframe with genes and their expression changes for the current contrast
  temp_df <- data.frame(gene=rownames(contrast_results[[i]]), 
                        expressionChange=contrast_results[[i]]$expressionChange)
  colnames(temp_df)[2] <- contrast_name  # Rename the second column to the current contrast name
  # Merge the temporary dataframe with the combined_results dataframe
  combined_results <- merge(combined_results, temp_df, by="gene", all=TRUE)
}

# Concatenate expression patterns to make model vector
combined_results$modelVector <- apply(combined_results[, -1], 1, 
                                      function(x) paste(x, collapse = ""))

# for Quaternary tree structure ###############################################################
combined_results$primary <- combined_results[,2]
combined_results$secondary <- paste0(combined_results[,3], combined_results[,4])
combined_results$tertiary <- paste0(combined_results[,5], combined_results[,6], combined_results[,7])
combined_results$quaternary <- paste0(combined_results[,8], combined_results[,9], combined_results[,10], combined_results[,11])



```

```{r, echo=FALSE, message=FALSE, results='asis'}
# Use the fully qualified function name to avoid conflicts
relevant_data <- combined_results %>%
  dplyr::select(starts_with("v to"), starts_with("mp to"), starts_with("lp to"), starts_with("el to"))

# Reshape data from wide to long format
long_data <- relevant_data %>%
  pivot_longer(cols = everything(), names_to = "transition", values_to = "event") %>%
  group_by(transition, event) %>%
  summarise(count = n(), .groups = 'drop')

# Spread the data back to wide format to see counts for each event type in each transition
wide_data <- long_data %>%
  pivot_wider(names_from = event, values_from = count, values_fill = list(count = 0))

# View the resulting DataFrame
print(wide_data)

wide <- as.data.frame(wide_data)

cat("
<style>
table caption {
  font-size: 16pt;
  font-weight: bold;
}
</style>
")

# Assuming 'samples' is another data frame you want to display
kable(wide, caption = "") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = FALSE)
```

```{r, echo=FALSE, message=FALSE, results='asis'}
selected_columns <- combined_results[, c(1, 12:16)]
head(selected_columns)
# Custom CSS to enlarge table captions
cat("
<style>
table caption {
  font-size: 16pt;
  font-weight: bold;
}
</style>
")

# Assuming 'samples' is another data frame you want to display
kable(head(combined_results[,1:11]), caption = "") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = FALSE)

kable(head(selected_columns), caption = "") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = FALSE)


```

## Mapping Specific Motif

```{r}
motif_index <- list()
x <- combined_results
vector <- paste0(x$`v to mp`,
                 x$`v to lp`,
                 x$`mp to el`,
                 x$`lp to el`,
                 x$`mp to l`,
                 x$`lp to l`,
                 x$`el to l`)

# Loop through each unique modelVector
for(model in unique(vector)) {
  # Subset the genes that match the current modelVector
  genes <- combined_results$gene[vector == model]
  # Store
  motif_index[[model]] <- genes
}

```

```{r, message=F, echo=F}
mapped_motif_index <- list()
for(model in names(motif_index)) {
  ensembl_ids <- motif_index[[model]]
  gene_symbols <- tryCatch({
    mapIds(org.Bt.eg.db,
           keys = ensembl_ids,
           column = "SYMBOL",
           keytype = "ENSEMBL",
           multivals = 'first')
  }, error = function(e) {
    message("Error with model ", model, ": ", e$message)
    return(NA)  # Return NA or any other placeholder that indicates an error
  })
  mapped_motif_index[[model]] <- gene_symbols
}

```

```{r, input_ui, echo=FALSE}
textInput("gene_input", "Type a Gene:", value = "LAMTOR1")
```

```{r, message=FALSE, echo=FALSE }

ncdf <- normalized_counts_df
ncdf$modelVector <- NULL

original_ensembl_ids <- row.names(ncdf)

# Map
gene_symbols <- list()
gene_symbols <- mapIds(org.Bt.eg.db,
                       keys = original_ensembl_ids,
                       column = "SYMBOL",
                       keytype = "ENSEMBL",
                       multivals = 'first')


unique_gene_symbols <- gene_symbols
duplicates <- duplicated(gene_symbols) | duplicated(gene_symbols, fromLast = TRUE)
unique_gene_symbols[duplicates] <- paste(gene_symbols[duplicates], original_ensembl_ids[duplicates], sep = "_")

# Replace NAs with the original Ensembl ID (if any NAs remain)
unique_gene_symbols[is.na(unique_gene_symbols)] <- original_ensembl_ids[is.na(unique_gene_symbols)]

# Update row names in ncdf with the unique gene symbols
row.names(ncdf) <- unique_gene_symbols
ncdf$gene <- rownames(ncdf)


myColors <- c(
  "v" = "#F8766D",
  "mp" = "#00BA38",
  "lp" = "#619CFF",
  "el" = "#F564E3", 
  "l" = "#FFC000" 
)


# Pivot the data to a long format
ld_goi <- pivot_longer(ncdf, cols = -gene, names_to = "group", values_to = "value")

# Clean group names by removing numeric suffixes
ld_goi$group <- gsub("(_\\d+)", "", ld_goi$group)



output$genePlot <- renderPlot({
  goi <- input$gene_input
  goi_model_vector <- ""

  # Check if the typed gene is in any model
  if(any(sapply(mapped_motif_index, function(x) goi %in% x))) {
    for(model in names(mapped_motif_index)) {
      if(goi %in% mapped_motif_index[[model]]) {
        goi_model_vector <- model
        break
      }
    }
  } else {
    # Optional: Display a message if the gene is not found (or handle it in another way)
    return(ggplot() + 
             labs(title = "Gene not found", subtitle = "Please check the gene name and try again."))
  }
  
  ld_gene_of_interest <- ld_goi %>% 
    filter(gene == goi) %>%
    filter(complete.cases(.)) %>%
    mutate(group = factor(group, levels = c("v", "mp", "lp", "el", "l")))
  
  ggplot(ld_gene_of_interest, aes(x = group, y = log10(value), fill = group)) +
    geom_boxplot() +
    theme_minimal() +
    geom_point(color="black", size=2, position = position_dodge(width = 0.75)) +
    geom_smooth(aes(group = 1), method = "loess", color = "navy", se = FALSE, size = 2.25) +
    stat_summary(fun = mean, geom = "point", color = "red", size = 3, shape = 18) +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    scale_x_discrete(labels = c("V", "MP", "LP", "EL", "L")) +
    labs(title = paste("Expression of", goi, "|", goi_model_vector),
         x = "Time Point",
         y = bquote(~Log[10]~ 'Normalized Counts')) +
    scale_fill_manual(values = myColors)
})
```

```{r, echo=FALSE}
plotOutput("genePlot")
```
