---
title: "Transcriptional Dynamics of Milk Produiction in Bovine Mammary Glands"
author: "Benji Lamp"
date: "4/24/24"
output: 
  html_document: 
    toc: true
    toc_float: true
    theme: united
    highlights: tango
---

# Introduction

It is understood that the mammary gland in cattle undergoes extensive morphological and physiological transformations to facilitate the production of milk. These changes are regulated at the genetic and epigenetic level, with specific patterns of expression driving maturation of the gland. This study aims to uncover the temporal expression patterns of mRNA in bovine mammary glands by performing timewise differential expression of sequencing data, then clustering based on temporal expression pattern.

# Setup {.tabset}

### Libraries

```{r, message=FALSE, warning=FALSE}
library(DESeq2)
library(dplyr)
library(AnnotationDbi)
library(org.Bt.eg.db)
library(sva)
library(readxl)
library(tidyr)
library(knitr)
library(kableExtra)
```

### Data

Import raw count files from github

```{r}
file_path <- 'https://raw.githubusercontent.com/blamp23/Cattle-lac-proj/main/Raw_data/Raw_Cattle_Counts.tabular'
df <- read.delim(file_path, header = TRUE, row.names = 1)
df <- df[-1]
```

# Filtering/loading {.tabset}

## Metadata

Creation of a dataframe that holds information about the experiment, as well as the batch effects

```{r, echo=FALSE}
file_path <- 'https://raw.githubusercontent.com/blamp23/Cattle-lac-proj/main/Raw_data/Raw_Cattle_Counts.tabular'
df <- read.delim(file_path, header = TRUE, row.names = 1)
df <- df[-1]


samples <- data.frame(
  row.names = c("v_468", "v_502", "v_504", "v_507", "v_509", "v_598", "v_610", 
                "mp_468", "mp_502", "mp_504", "mp_507", "mp_509", "mp_598", "mp_610",
                "lp_468", "lp_502", "lp_504", "lp_509", "lp_507", "lp_598", "lp_610",
                "el_468", "el_502", "el_504", "el_507", "el_509", "el_598", "el_610",
                "l_468", "l_502", "l_504", "l_507", "l_509", "l_598", "l_610"),
  condition = c(rep("v", 7), rep("mp", 7), rep("lp", 7), rep("el", 7), rep("l", 7)),
  animal = c(468, 502, 504, 507, 509, 598, 610,
             468, 502, 504, 507, 509, 598, 610,
             468, 502, 504, 509, 507, 598, 610,
             468, 502, 504, 507, 509, 598, 610,
             468, 502, 504, 507, 509, 598, 610),
  batch = c(1, 1, 1, 1, 1, 1, 1, 
            0, 1, 0, 1, 0, 0, 0,
            0, 1, 0, 0, 1, 0, 0,
            0, 1, 0, 1, 0, 0, 0, 
            1, 1, 1, 1, 1, 1, 1 ),
  BCM = c(1, 1, 1, 1, 1, 1, 1,
          0, 0, 0, 0, 0, 0, 0,
          0, 0, 0, 0, 0, 0, 0,
          0, 0, 0, 0, 0, 0, 0,
          1, 1, 1, 1, 1, 1, 1)
)

new_colnames <- c(row.names(samples))
colnames(df) <- new_colnames

colnames(df)[28] <- "lp_610"  # lp_610 and el_610 are swapped samples
colnames(df)[21] <- "el_610"

el_610 <- which(names(df) == "el_610")
lp_610 <- which(names(df) == "lp_610")

# Create a new vector of column names with the positions of these two columns swapped
new_order <- names(df)
new_order[c(el_610, lp_610)] <- new_order[c(lp_610, el_610)]

# Reorder the columns in the DataFrame
df <- df[, new_order]

samples <- samples[colnames(df), ]

print(samples)
head(df)

```

## Filtering

```{r}
# Filter Method A ##############################################################
# originaly 27607 genes
# filtered to 17344 genes
# Step 1: Splitting Data by Condition
conditions <- unique(samples$condition)
split_data <- lapply(conditions, function(cond) {
  df[, samples$condition == cond]
})

# Step 2: Define Filtering Function
count_high_expressions <- function(data) {
  apply(data, 1, function(gene) {
    sum(gene >= 3) >= 4  # Returns TRUE if at least 4 out of 7 samples have counts â‰¥ 3
  })
}
high_exp_counts <- lapply(split_data, count_high_expressions)

# Step 3: Aggregate Results for inclusion
gene_inclusion_filter <- rowSums(do.call(cbind, high_exp_counts)) >= 1 # If 1/5 timepoints meet crit, gene is included

# Step 4: Filter Genes
df <- df[gene_inclusion_filter, ]
```

## Filtered Counts

```{r, echo=FALSE, results='asis'}
library(knitr)
library(kableExtra)

# Custom CSS to enlarge table captions
cat("
<style>
table caption {
  font-size: 16pt;
  font-weight: bold;
}
</style>
")

# Assuming 'samples' is another data frame you want to display
kable(samples, caption = "Sample Data") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = FALSE)

# Displaying the first few rows of df
df_head <- head(df)

# Splitting df into two halves if it has many columns
half_col <- ceiling(ncol(df_head) / 2)
df_head_first_part <- df_head[, 1:half_col]
df_head_second_part <- df_head[, (half_col+1):ncol(df_head)]

# Display the first half
kable(df_head_first_part, caption = "Filtered Counts") %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = FALSE)

# Display the second half
kable(df_head_second_part) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed"), full_width = FALSE)
```

# Norm {.tabset}

## ComBatSeq

```{r}
count_matrix <- as.matrix(df)
cdf <- ComBat_seq(count_matrix, batch = samples$batch, group = samples$condition)
```

## Normilisation

```{r}
dds <- DESeqDataSetFromMatrix(
  countData = cdf,
  colData = samples,
  design = ~ condition
)


dds <- DESeq(dds)
res <- results(dds)


dds_count <- estimateSizeFactors(dds)
normalized_counts <- counts(dds_count, normalized=TRUE)
```

# Timecourse {.tabset}

## Averages DF

```{r}
# Time-point Averages of Normalized Counts ######################################
normalized_counts_df <- as.data.frame(normalized_counts)
# Define/assign column names
col_names <- c(rep("v", 7), rep("mp", 7), rep("lp", 7), rep("el", 7), rep("l", 7))
names(normalized_counts_df) <- col_names

# Prepare an empty data frame for storing averages, with correct dimensions and column names
averages_df <- data.frame(matrix(ncol = length(unique(col_names)), nrow = nrow(normalized_counts_df)))
names(averages_df) <- unique(col_names)
rownames(averages_df) <- rownames(normalized_counts_df)

# Loop through each time-point to calculate and store row means
for (timepoint in unique(col_names)) {
  # Identify columns belonging to the current time-point
  cols <- which(col_names == timepoint)
  # Calculate and store
  averages_df[[timepoint]] <- rowMeans(normalized_counts_df[, cols], na.rm = TRUE)
}
head(averages_df)
```

## Pairwise Comparison

```{r}
contrasts <- list(
  c("condition", "l", "v"),
  #
  c("condition", "el", "v"),
  c("condition", "l", "mp"),
  #
  c("condition", "lp", "v"),
  c("condition", "el", "mp"),
  c("condition", "l", "lp"),
  #
  c("condition", "mp", "v"),
  c("condition", "lp", "mp"),
  c("condition", "el", "lp"),
  c("condition", "l", "el")
)

contrast_results <- list()

# DE for pairs in contrasts
for (i in seq_along(contrasts)) {
  contrast = contrasts[[i]]
  res <- results(dds, contrast=contrast)
  
  # Determine if the mean expression increases or decreases
  condition1_mean <- averages_df[[contrast[3]]]  # Mean for the first condition
  condition2_mean <- averages_df[[contrast[2]]]  # Mean for the second condition
  expressionDirection <- ifelse(condition1_mean < condition2_mean, "I", "D")
  
  # Hypothesis test, FTR null assign 'S', Rej null assign I/D
  res$expressionChange <- ifelse(res$padj > 0.05, "S", expressionDirection)
  
  contrast_results[[paste(contrast[3], "to", contrast[2])]] <- res
}
```

## Motif Structuring

```{r}
combined_results <- data.frame(gene=rownames(contrast_results[[1]]))
# Loop through each contrast result to combine them
for (i in seq_along(contrast_results)) {
  contrast_name <- names(contrast_results)[i]  # Get the name of the current contrast
  # Create a temporary dataframe with genes and their expression changes for the current contrast
  temp_df <- data.frame(gene=rownames(contrast_results[[i]]), 
                        expressionChange=contrast_results[[i]]$expressionChange)
  colnames(temp_df)[2] <- contrast_name  # Rename the second column to the current contrast name
  # Merge the temporary dataframe with the combined_results dataframe
  combined_results <- merge(combined_results, temp_df, by="gene", all=TRUE)
}

# Concatenate expression patterns to make model vector
combined_results$modelVector <- apply(combined_results[, -1], 1, 
                                      function(x) paste(x, collapse = ""))

# for Quaternary tree structure ###############################################################
combined_results$primary <- combined_results[,2]
combined_results$secondary <- paste0(combined_results[,3], combined_results[,4])
combined_results$tertiary <- paste0(combined_results[,5], combined_results[,6], combined_results[,7])
combined_results$quaternary <- paste0(combined_results[,8], combined_results[,9], combined_results[,10], combined_results[,11])

head(combined_results)

```

## Mapping Specific Motif 

```{r}
motif_index <- list()
x <- combined_results
vector <- paste0(x$`v to mp`,
                 x$`v to lp`,
                 x$`mp to el`,
                 x$`lp to el`,
                 x$`mp to l`,
                 x$`lp to l`,
                 x$`el to l`)

# Loop through each unique modelVector
for(model in unique(vector)) {
  # Subset the genes that match the current modelVector
  genes <- combined_results$gene[vector == model]
  # Store
  motif_index[[model]] <- genes
}
summary(motif_index)
mi_mrna <- motif_index
```
